// @salutejs/plasma-web/components/Calendar/CalendarMonths.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CalendarMonths = void 0;

var _react = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("react"));

var _styledComponents = /*#__PURE__*/_interopRequireDefault( /*#__PURE__*/require("styled-components"));

var _plasmaTypo = /*#__PURE__*/require("@salutejs/plasma-typo");

var _hooks = /*#__PURE__*/require("./hooks");

var _mixins = /*#__PURE__*/require("./mixins");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]; if (_i == null) return; var _arr = []; var _n = true; var _d = false; var _s, _e; try { for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

var StyledCalendarMonths = /*#__PURE__*/_styledComponents["default"].div.withConfig({
  componentId: "plasma-web__sc-p75yqi-0"
})(["padding:0.5rem 1.5rem 1.5rem;box-sizing:border-box;"]);

var StyledFlex = /*#__PURE__*/_styledComponents["default"].div.withConfig({
  componentId: "plasma-web__sc-p75yqi-1"
})(["", ";"], _mixins.flexCenter);

var StyledMonth = /*#__PURE__*/_styledComponents["default"].div.withConfig({
  componentId: "plasma-web__sc-p75yqi-2"
})(["border-radius:0.438rem;", ";"], _mixins.flexCenter);

var StyledMonthRoot = /*#__PURE__*/_styledComponents["default"].div.withConfig({
  componentId: "plasma-web__sc-p75yqi-3"
})(["", ";position:relative;box-sizing:border-box;min-height:3.5rem;border-radius:0.5rem;flex:1;", ";", ";"], _plasmaTypo.bodyS, _mixins.flexCenter, function (_ref) {
  var isSelected = _ref.isSelected,
      isCurrent = _ref.isCurrent;
  return (0, _mixins.selected)({
    StyledItem: StyledMonth,
    minWidth: 5.25,
    minHeight: 3.25,
    isSelected: isSelected,
    isCurrent: isCurrent
  });
});
/**
 * Компонент месяцев в календаре.
 */


var CalendarMonths = function CalendarMonths(_ref2) {
  var currentDate = _ref2.date,
      selectIndexes = _ref2.selectIndexes,
      outerRefs = _ref2.outerRefs,
      onChangeMonth = _ref2.onChangeMonth,
      onSetSelected = _ref2.onSetSelected,
      onKeyDown = _ref2.onKeyDown;

  var _useMonths = (0, _hooks.useMonths)(currentDate),
      _useMonths2 = _slicedToArray(_useMonths, 2),
      months = _useMonths2[0],
      selected = _useMonths2[1];

  var selectedRef = (0, _react.useRef)(selected);
  var onSetSelectedRef = (0, _react.useRef)(onSetSelected);
  var handleOnChangeMonth = (0, _react.useCallback)(function (event) {
    /**
     * нужно вызвать stopImmediatePropagation для случаев, когда
     * обработчик события onClick навешивается снаружи.
     * Как, например, в компоненте Popup
     */
    event.nativeEvent.stopImmediatePropagation();
    var monthIndex = event.currentTarget.dataset.monthIndex;
    onChangeMonth(Number(monthIndex));
  }, [onChangeMonth]);
  var getRefs = (0, _react.useCallback)(function (element, i, j) {
    outerRefs.current[i][j] = element;
  }, [outerRefs]);
  (0, _react.useEffect)(function () {
    if (selectedRef.current) {
      var _onSetSelectedRef$cur;

      (_onSetSelectedRef$cur = onSetSelectedRef.current) === null || _onSetSelectedRef$cur === void 0 ? void 0 : _onSetSelectedRef$cur.call(onSetSelectedRef, selectedRef.current);
    }
  }, []);
  return /*#__PURE__*/_react["default"].createElement(StyledCalendarMonths, {
    role: "grid",
    "aria-labelledby": "id-grid-label",
    onKeyDown: onKeyDown
  }, months.map(function (month, i) {
    return /*#__PURE__*/_react["default"].createElement(StyledFlex, {
      role: "row",
      key: i
    }, month.map(function (_ref3, j) {
      var monthName = _ref3.monthName,
          monthIndex = _ref3.monthIndex,
          isSelected = _ref3.isSelected,
          isCurrent = _ref3.isCurrent,
          monthFullName = _ref3.monthFullName;
      return /*#__PURE__*/_react["default"].createElement(StyledMonthRoot, {
        id: "month-test-".concat(i, "-").concat(j),
        ref: function ref(element) {
          return getRefs(element, i, j);
        },
        tabIndex: i === (selectIndexes === null || selectIndexes === void 0 ? void 0 : selectIndexes[0]) && j === (selectIndexes === null || selectIndexes === void 0 ? void 0 : selectIndexes[1]) ? 0 : -1,
        isCurrent: isCurrent,
        isSelected: isSelected,
        onClick: handleOnChangeMonth,
        "data-month-index": monthIndex,
        "aria-selected": isSelected,
        role: "gridcell",
        key: "StyledMonth-".concat(i, "-").concat(j),
        "aria-label": monthFullName
      }, /*#__PURE__*/_react["default"].createElement(StyledMonth, null, monthName));
    }));
  }));
};

exports.CalendarMonths = CalendarMonths;
