// @salutejs/plasma-web/es/components/Calendar/CalendarYears.js
function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]; if (_i == null) return; var _arr = []; var _n = true; var _d = false; var _s, _e; try { for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

import React, { useCallback, useEffect, useRef } from 'react';
import styled from 'styled-components';
import { bodyS } from '@salutejs/plasma-typo';
import { useYears } from './hooks';
import { flexCenter, selected as selectedMixin } from './mixins';
var StyledCalendarYears = /*#__PURE__*/styled.div.withConfig({
  componentId: "plasma-web__sc-d58fi6-0"
})(["padding:0.5rem 1.5rem 1.5rem;box-sizing:border-box;"]);
var StyledFlex = /*#__PURE__*/styled.div.withConfig({
  componentId: "plasma-web__sc-d58fi6-1"
})(["", ";"], flexCenter);
var StyledYear = /*#__PURE__*/styled.div.withConfig({
  componentId: "plasma-web__sc-d58fi6-2"
})(["border-radius:0.438rem;", ";"], flexCenter);
var StyledYearRoot = /*#__PURE__*/styled.div.withConfig({
  componentId: "plasma-web__sc-d58fi6-3"
})(["", ";position:relative;box-sizing:border-box;min-height:3.5rem;border-radius:0.5rem;flex:1;", ";", ";"], bodyS, flexCenter, function (_ref) {
  var isSelected = _ref.isSelected,
      isCurrent = _ref.isCurrent;
  return selectedMixin({
    StyledItem: StyledYear,
    minWidth: 5.25,
    minHeight: 3.25,
    isSelected: isSelected,
    isCurrent: isCurrent
  });
});
/**
 * Компонент годов в календаре.
 */

export var CalendarYears = function CalendarYears(_ref2) {
  var currentDate = _ref2.date,
      startYear = _ref2.startYear,
      selectIndexes = _ref2.selectIndexes,
      outerRefs = _ref2.outerRefs,
      onChangeYear = _ref2.onChangeYear,
      onSetSelected = _ref2.onSetSelected,
      onKeyDown = _ref2.onKeyDown;

  var _useYears = useYears(currentDate, startYear),
      _useYears2 = _slicedToArray(_useYears, 2),
      years = _useYears2[0],
      selected = _useYears2[1];

  var selectedRef = useRef(selected);
  var onSetSelectedRef = useRef(onSetSelected);
  var handleOnChangeYear = useCallback(function (event) {
    /**
     * нужно вызвать stopImmediatePropagation для случаев, когда
     * обработчик события onClick навешивается снаружи.
     * Как, например, в компоненте Popup
     */
    event.nativeEvent.stopImmediatePropagation();
    var year = event.currentTarget.dataset.year;
    onChangeYear(Number(year));
  }, [onChangeYear]);
  var getRefs = useCallback(function (element, i, j) {
    outerRefs.current[i][j] = element;
  }, [outerRefs]);
  useEffect(function () {
    if (selectedRef.current) {
      var _onSetSelectedRef$cur;

      (_onSetSelectedRef$cur = onSetSelectedRef.current) === null || _onSetSelectedRef$cur === void 0 ? void 0 : _onSetSelectedRef$cur.call(onSetSelectedRef, selectedRef.current);
    }
  }, []);
  return /*#__PURE__*/React.createElement(StyledCalendarYears, {
    role: "grid",
    "aria-labelledby": "id-grid-label",
    onKeyDown: onKeyDown
  }, years.map(function (year, i) {
    return /*#__PURE__*/React.createElement(StyledFlex, {
      role: "row",
      key: i
    }, year.map(function (_ref3, j) {
      var yearValue = _ref3.yearValue,
          isSelected = _ref3.isSelected,
          isCurrent = _ref3.isCurrent;
      return /*#__PURE__*/React.createElement(StyledYearRoot, {
        ref: function ref(element) {
          return getRefs(element, i, j);
        },
        tabIndex: i === (selectIndexes === null || selectIndexes === void 0 ? void 0 : selectIndexes[0]) && j === (selectIndexes === null || selectIndexes === void 0 ? void 0 : selectIndexes[1]) ? 0 : -1,
        isCurrent: isCurrent,
        isSelected: isSelected,
        onClick: handleOnChangeYear,
        "data-year": yearValue,
        "aria-selected": isSelected,
        role: "gridcell",
        key: "StyledYear-".concat(i, "-").concat(j),
        "aria-label": String(yearValue)
      }, /*#__PURE__*/React.createElement(StyledYear, null, yearValue));
    }));
  }));
};
