// @salutejs/plasma-web/es/components/Calendar/CalendarBase.js
var _excluded = ["value", "min", "max", "type", "eventList", "disabledList", "onChangeValue"];

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]; if (_i == null) return; var _arr = []; var _n = true; var _d = false; var _s, _e; try { for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

import React, { useCallback, useMemo, useReducer, useState } from 'react';
import styled from 'styled-components';
import { CalendarState } from './types';
import { isValueUpdate, YEAR_RENDER_COUNT } from './utils';
import { CalendarDays } from './CalendarDays';
import { CalendarMonths } from './CalendarMonths';
import { CalendarHeader } from './CalendarHeader';
import { CalendarYears } from './CalendarYears';
import { useKeyNavigation } from './hooks';
import { getInitialState, reducer } from './store/reducer';
import { ActionType } from './store/types';
var StyledCalendar = /*#__PURE__*/styled.div.withConfig({
  componentId: "plasma-web__sc-1i4r6fa-0"
})(["position:relative;user-select:none;z-index:1;width:19.5rem;height:19.5rem;"]);
var IsOutOfRange = /*#__PURE__*/styled.div.withConfig({
  componentId: "plasma-web__sc-1i4r6fa-1"
})(["position:absolute;padding:0;margin:0;height:0;width:0;border:0;overflow:hidden;clip:rect(0 0 0 0);"]);
/**
 * Компонент календаря.
 */

export var CalendarBase = function CalendarBase(_ref) {
  var externalValue = _ref.value,
      min = _ref.min,
      max = _ref.max,
      _ref$type = _ref.type,
      type = _ref$type === void 0 ? 'Days' : _ref$type,
      eventList = _ref.eventList,
      disabledList = _ref.disabledList,
      onChangeValue = _ref.onChangeValue,
      rest = _objectWithoutProperties(_ref, _excluded);

  var _useMemo = useMemo(function () {
    return Array.isArray(externalValue) ? externalValue : [externalValue];
  }, [externalValue]),
      _useMemo2 = _slicedToArray(_useMemo, 2),
      firstValue = _useMemo2[0],
      secondValue = _useMemo2[1];

  var value = secondValue || firstValue;

  var _useState = useState(),
      _useState2 = _slicedToArray(_useState, 2),
      hoveredDay = _useState2[0],
      setHoveredDay = _useState2[1];

  var _useState3 = useState(type),
      _useState4 = _slicedToArray(_useState3, 2),
      prevType = _useState4[0],
      setPrevType = _useState4[1];

  var _useState5 = useState(value),
      _useState6 = _slicedToArray(_useState5, 2),
      prevValue = _useState6[0],
      setPrevValue = _useState6[1];

  var _useState7 = useState(0),
      _useState8 = _slicedToArray(_useState7, 2),
      outOfRangeKey = _useState8[0],
      setOutOfRangeKey = _useState8[1];

  var _useReducer = useReducer(reducer, getInitialState(value, [5, 6], type)),
      _useReducer2 = _slicedToArray(_useReducer, 2),
      state = _useReducer2[0],
      dispatch = _useReducer2[1];

  var date = state.date,
      calendarState = state.calendarState,
      startYear = state.startYear,
      size = state.size;
  var handlePrev = useCallback(function () {
    var withShift = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : false;

    if (calendarState === CalendarState.Days) {
      if (withShift) {
        dispatch({
          type: ActionType.PREVIOUS_YEAR,
          payload: {
            step: 1
          }
        });
        return;
      }

      dispatch({
        type: ActionType.PREVIOUS_MONTH,
        payload: {
          monthIndex: date.monthIndex,
          year: date.year
        }
      });
      return;
    }

    if (calendarState === CalendarState.Months) {
      dispatch({
        type: ActionType.PREVIOUS_YEAR,
        payload: {
          step: 1
        }
      });
      return;
    }

    if (calendarState === CalendarState.Years) {
      dispatch({
        type: ActionType.PREVIOUS_START_YEAR,
        payload: {
          yearsCount: YEAR_RENDER_COUNT
        }
      });
    }
  }, [date, calendarState]);
  var handleNext = useCallback(function () {
    var withShift = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : false;

    if (calendarState === CalendarState.Days) {
      if (withShift) {
        dispatch({
          type: ActionType.NEXT_YEAR,
          payload: {
            step: 1
          }
        });
        return;
      }

      dispatch({
        type: ActionType.NEXT_MONTH,
        payload: {
          monthIndex: date.monthIndex,
          year: date.year
        }
      });
      return;
    }

    if (calendarState === CalendarState.Months) {
      dispatch({
        type: ActionType.NEXT_YEAR,
        payload: {
          step: 1
        }
      });
      return;
    }

    if (calendarState === CalendarState.Years) {
      dispatch({
        type: ActionType.NEXT_START_YEAR,
        payload: {
          yearsCount: YEAR_RENDER_COUNT
        }
      });
    }
  }, [date, calendarState]);

  var _useKeyNavigation = useKeyNavigation({
    size: size,
    onNext: handleNext,
    onPrev: handlePrev
  }),
      _useKeyNavigation2 = _slicedToArray(_useKeyNavigation, 5),
      selectIndexes = _useKeyNavigation2[0],
      onKeyDown = _useKeyNavigation2[1],
      onSelectIndexes = _useKeyNavigation2[2],
      outerRefs = _useKeyNavigation2[3],
      isOutOfRange = _useKeyNavigation2[4];

  var handleOnChangeDay = useCallback(function (newDate, coord) {
    var newDay = new Date(newDate.year, newDate.monthIndex, newDate.day);
    onChangeValue(newDay);
    onSelectIndexes(coord);
  }, [onChangeValue, onSelectIndexes]);
  var handleOnChangeMonth = useCallback(function (monthIndex) {
    dispatch({
      type: ActionType.UPDATE_MONTH,
      payload: {
        calendarState: CalendarState.Days,
        monthIndex: monthIndex,
        size: [5, 6]
      }
    });
  }, []);
  var handleOnChangeYear = useCallback(function (year) {
    dispatch({
      type: ActionType.UPDATE_YEAR,
      payload: {
        calendarState: CalendarState.Months,
        year: year
      }
    });
  }, []);
  var handleUpdateCalendarState = useCallback(function (newCalendarState, newSize) {
    dispatch({
      type: ActionType.UPDATE_CALENDAR_STATE,
      payload: {
        calendarState: newCalendarState,
        size: newSize
      }
    });
  }, []);

  if (value && prevValue && isValueUpdate(value, prevValue)) {
    dispatch({
      type: ActionType.UPDATE_DATE,
      payload: {
        value: value
      }
    });
    setPrevValue(value);
  }

  if (prevType !== type) {
    dispatch({
      type: ActionType.UPDATE_CALENDAR_STATE,
      payload: {
        calendarState: type
      }
    });
    setPrevType(type);
  } // Изменяем ключ каждый раз как пытаемся перейти на даты которые находятся за пределами min/max ограничений.
  // Это необходимо для того чтобы screen-reader корректно озвучивал уведомление aria-live="assertive"
  // о том что нет доступных дат


  var handleKeyDown = useCallback(function (event) {
    setOutOfRangeKey(function (previousState) {
      return Number(!previousState);
    });
    onKeyDown(event);
  }, [onKeyDown]);
  return /*#__PURE__*/React.createElement(StyledCalendar, _extends({
    "aria-label": "\u0412\u044B\u0431\u043E\u0440 \u0434\u0430\u0442\u044B"
  }, rest), isOutOfRange && /*#__PURE__*/React.createElement(IsOutOfRange, {
    key: outOfRangeKey,
    "aria-atomic": "true",
    role: "alert",
    "aria-live": "assertive",
    "aria-relevant": "additions"
  }, "\u0414\u0430\u043B\u0435\u0435 \u043D\u0435\u0442 \u0434\u043E\u0441\u0442\u0443\u043F\u043D\u044B\u0445 \u0434\u0430\u0442."), /*#__PURE__*/React.createElement(CalendarHeader, {
    firstDate: date,
    startYear: startYear,
    type: calendarState,
    onPrev: handlePrev,
    onNext: handleNext,
    onUpdateCalendarState: handleUpdateCalendarState
  }), calendarState === CalendarState.Days && /*#__PURE__*/React.createElement(CalendarDays, {
    eventList: eventList,
    disabledList: disabledList,
    min: min,
    max: max,
    value: externalValue,
    date: date,
    hoveredDay: hoveredDay,
    selectIndexes: selectIndexes,
    onChangeDay: handleOnChangeDay,
    onHoverDay: setHoveredDay,
    onSetSelected: onSelectIndexes,
    onKeyDown: handleKeyDown,
    outerRefs: outerRefs
  }), calendarState === CalendarState.Months && /*#__PURE__*/React.createElement(CalendarMonths, {
    date: date,
    selectIndexes: selectIndexes,
    onChangeMonth: handleOnChangeMonth,
    onSetSelected: onSelectIndexes,
    onKeyDown: onKeyDown,
    outerRefs: outerRefs
  }), calendarState === CalendarState.Years && /*#__PURE__*/React.createElement(CalendarYears, {
    date: date,
    startYear: startYear,
    selectIndexes: selectIndexes,
    onChangeYear: handleOnChangeYear,
    onSetSelected: onSelectIndexes,
    onKeyDown: onKeyDown,
    outerRefs: outerRefs
  }));
};
