// @salutejs/plasma-web/es/components/Calendar/CalendarDayItem.js
var _excluded = ["isFocused", "dayOfWeek", "disabled", "isCurrent", "isSelected", "isDayInCurrentMonth", "isDouble", "inRange", "isHovered", "sideInRange", "eventList", "day", "monthIndex", "year", "onClick", "onMouseOver", "onFocus", "disabledArrowKey", "disabledMonths"];

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

import React, { forwardRef, memo } from 'react';
import styled, { css } from 'styled-components';
import { accent, tertiary, secondary, primary, surfaceLiquid02, addFocus } from '@salutejs/plasma-core';
import { bodyS } from '@salutejs/plasma-typo';
import { flexCenter, selected } from './mixins';
var StyledDay = /*#__PURE__*/styled.div.withConfig({
  componentId: "plasma-web__sc-1voj0go-0"
})(["border-radius:0.375rem;align-items:center;", ";", ""], flexCenter, function (_ref) {
  var inRange = _ref.inRange;
  return inRange && css(["&::before{content:'';z-index:-1;position:absolute;width:2.5rem;height:1.875rem;background:", ";}"], surfaceLiquid02);
});

var setSide = function setSide(side, isCurrent, isSelected) {
  switch (side) {
    case 'left':
      return "left: ".concat(!isSelected && isCurrent ? '-1px' : '0');

    case 'right':
      return "right: ".concat(!isSelected && isCurrent ? '-1px' : '0');

    default:
      return undefined;
  }
};

var StyledDayRoot = /*#__PURE__*/styled.div.withConfig({
  componentId: "plasma-web__sc-1voj0go-1"
})(["", ";position:relative;box-sizing:border-box;min-width:2.5rem;min-height:2rem;border-radius:0.5rem;", ";", " ", " ", " ", " ", ";", ""], bodyS, flexCenter, function (_ref2) {
  var isDayInCurrentMonth = _ref2.isDayInCurrentMonth;
  return css(["color:", ";"], isDayInCurrentMonth ? primary : secondary);
}, function (_ref3) {
  var isDayInCurrentMonth = _ref3.isDayInCurrentMonth,
      isDouble = _ref3.isDouble;
  return css(["visibility:", ";"], !isDayInCurrentMonth && isDouble ? 'hidden' : 'visible');
}, function (_ref4) {
  var sideInRange = _ref4.sideInRange,
      isCurrent = _ref4.isCurrent,
      isSelected = _ref4.isSelected;
  return sideInRange && css(["", "::before{content:'';z-index:-1;position:absolute;width:0.313rem;height:1.875rem;background:", ";", ";}"], StyledDay, surfaceLiquid02, setSide(sideInRange, isCurrent, isSelected));
}, function (_ref5) {
  var dayOfWeek = _ref5.dayOfWeek;
  return dayOfWeek && css(["color:", ";"], tertiary);
}, function (_ref6) {
  var isSelected = _ref6.isSelected,
      isCurrent = _ref6.isCurrent,
      isHovered = _ref6.isHovered,
      dayOfWeek = _ref6.dayOfWeek,
      $disabled = _ref6.$disabled;
  return !dayOfWeek && !$disabled && selected({
    StyledItem: StyledDay,
    minWidth: 2.25,
    minHeight: 1.75,
    isSelected: isSelected,
    isCurrent: isCurrent,
    isHovered: isHovered
  });
}, function (_ref7) {
  var $disabled = _ref7.$disabled,
      isCurrent = _ref7.isCurrent;
  return $disabled && css(["cursor:not-allowed;opacity:0.4;", ";"], addFocus({
    outlineRadius: '0.563rem',
    outlineSize: '0.063rem',
    outlineOffset: isCurrent ? '0.125rem' : '0.063rem'
  }));
});
var StyledEvents = /*#__PURE__*/styled.div.withConfig({
  componentId: "plasma-web__sc-1voj0go-2"
})(["display:flex;position:absolute;bottom:0.25rem;"]);
var StyledEvent = /*#__PURE__*/styled.div.withConfig({
  componentId: "plasma-web__sc-1voj0go-3"
})(["margin:0 0.063rem;width:0.188rem;height:0.188rem;border-radius:50%;", ""], function (_ref8) {
  var _ref8$color = _ref8.color,
      color = _ref8$color === void 0 ? accent : _ref8$color;
  return css(["background-color:", ";"], color);
});
/**
 * Компонент дня в календаре.
 */

export var CalendarDayItem = /*#__PURE__*/memo( /*#__PURE__*/forwardRef(function (_ref9, outerRef) {
  var isFocused = _ref9.isFocused,
      dayOfWeek = _ref9.dayOfWeek,
      disabled = _ref9.disabled,
      isCurrent = _ref9.isCurrent,
      isSelected = _ref9.isSelected,
      isDayInCurrentMonth = _ref9.isDayInCurrentMonth,
      isDouble = _ref9.isDouble,
      inRange = _ref9.inRange,
      isHovered = _ref9.isHovered,
      sideInRange = _ref9.sideInRange,
      _ref9$eventList = _ref9.eventList,
      eventList = _ref9$eventList === void 0 ? [] : _ref9$eventList,
      day = _ref9.day,
      monthIndex = _ref9.monthIndex,
      year = _ref9.year,
      onClick = _ref9.onClick,
      onMouseOver = _ref9.onMouseOver,
      onFocus = _ref9.onFocus,
      disabledArrowKey = _ref9.disabledArrowKey,
      disabledMonths = _ref9.disabledMonths,
      rest = _objectWithoutProperties(_ref9, _excluded);

  return /*#__PURE__*/React.createElement(StyledDayRoot, _extends({
    ref: outerRef,
    tabIndex: isFocused ? 0 : -1,
    dayOfWeek: dayOfWeek,
    isCurrent: isCurrent,
    isSelected: isSelected,
    isDayInCurrentMonth: isDayInCurrentMonth,
    isDouble: isDouble,
    isHovered: isHovered,
    sideInRange: sideInRange,
    onClick: disabled ? undefined : onClick,
    onMouseOver: onMouseOver,
    onFocus: onFocus,
    "data-day": day,
    "data-month-index": monthIndex,
    "data-year": year,
    "data-disabled-arrow-key": disabledArrowKey || undefined,
    "data-disabled-months": disabledMonths || undefined,
    "aria-selected": isSelected,
    "aria-disabled": disabled,
    "aria-describedby": disabled ? 'withShift' : undefined,
    $disabled: disabled
  }, rest), /*#__PURE__*/React.createElement(StyledDay, {
    inRange: inRange
  }, day), /*#__PURE__*/React.createElement(StyledEvents, {
    "aria-hidden": true
  }, [eventList[0], eventList[1], eventList[2]].map(function (event, i) {
    return event && /*#__PURE__*/React.createElement(StyledEvent, _extends({
      key: "".concat(event.date, "-").concat(i)
    }, event));
  })));
}));
